File: src/main/java/com/galaxy/Restaurantinformationsystem/service/ReviewService.java
package com.galaxy.Restaurantinformationsystem.service;

import com.galaxy.Restaurantinformationsystem.dto.ReviewDTO;
import com.galaxy.Restaurantinformationsystem.dto.ReviewFormDTO;
import com.galaxy.Restaurantinformationsystem.entity.ReviewEntity;
import com.galaxy.Restaurantinformationsystem.entity.StoreEntity;
import com.galaxy.Restaurantinformationsystem.repository.MenuRepository;
import com.galaxy.Restaurantinformationsystem.repository.ReviewRepository;
import com.galaxy.Restaurantinformationsystem.repository.StoreRepository;
import com.galaxy.Restaurantinformationsystem.repository.UserRepository;
import lombok.AllArgsConstructor;
import org.modelmapper.ModelMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
public class ReviewService {

    ReviewRepository reviewRepository;
    MenuRepository menuRepository;
    UserRepository userRepository;
    StoreRepository storeRepository;
    ModelMapper modelMapper;


    public ReviewDTO createReview(Long storeId, ReviewFormDTO reviewFormDTO, String token) {
        ReviewEntity reviewEntity = modelMapper.map(reviewFormDTO, ReviewEntity.class);
        reviewEntity.setStoreEntity(StoreEntity.builder().id(storeId).build());
        // TODO: token으로 부터 id를 받아와서 userEntity에 넣어주기
        //reviewEntity.setUserEntity(UserEntity.builder().id().build());
        return modelMapper.map(reviewRepository.save(reviewEntity), ReviewDTO.class);
    }

    public ReviewDTO updateReview(Long reviewid, ReviewFormDTO reviewFormDTO, String token) {
        Optional<ReviewEntity> reviewEntity = reviewRepository.findById(reviewid);

        if (reviewEntity.isEmpty()) {
            throw new RuntimeException("리뷰가 존재하지 않습니다.");
        }
//         작성자 본인이 아니면 수정 불가
//        if (reviewEntity.get().getUserEntity().getId() != 1L){
//            throw new RuntimeException("권한이 없습니다.");
//        }

        reviewEntity.get().setTitle(reviewFormDTO.getTitle());
        reviewEntity.get().setContent(reviewFormDTO.getContent());
        reviewEntity.get().setImage(reviewFormDTO.getImage());
        reviewEntity.get().setScore(reviewFormDTO.getScore());
        return modelMapper.map(reviewRepository.save(reviewEntity.get()), ReviewDTO.class);

    }

    public List<ReviewDTO> getMyReview(Long userId) {
        // token에서 id 받아오기
        List<ReviewEntity> reviewEntityList = reviewRepository.findByUserEntity_Id(userId);

        List<ReviewDTO> reviewDTOList = new ArrayList<>();
        for (ReviewEntity reviewEntity : reviewEntityList) {
            reviewDTOList.add(modelMapper.map(reviewEntity, ReviewDTO.class));
        }

        return reviewDTOList;
    }

    @Transactional
    public void deleteReview(Long reviewId, String token) {
        Optional<ReviewEntity> reviewEntity = reviewRepository.findById(reviewId);

        if (reviewEntity.isEmpty()) {
            throw new RuntimeException("리뷰가 존재하지 않습니다.");
        }

        // 작성자 본인이 아니면 삭제 불가
//        if (reviewEntity.get().getUserEntity().getId() != 1L){
//            throw new RuntimeException("권한이 없습니다.");
//        }

        reviewRepository.delete(reviewEntity.get());
    }

}

File: src/main/java/com/galaxy/Restaurantinformationsystem/service/ReviewService.kt
package com.galaxy.Restaurantinformationsystem.service

import com.galaxy.Restaurantinformationsystem.dto.ReviewRequestDto
import com.galaxy.Restaurantinformationsystem.dto.ReviewResponseDto
import com.galaxy.Restaurantinformationsystem.entity.ReviewEntity
import com.galaxy.Restaurantinformationsystem.mapper.StoreMapper
import com.galaxy.Restaurantinformationsystem.repository.MenuRepository
import com.galaxy.Restaurantinformationsystem.repository.ReviewRepository
import com.galaxy.Restaurantinformationsystem.repository.StoreRepository
import com.galaxy.Restaurantinformationsystem.repository.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ReviewService(
    var reviewRepository: ReviewRepository,
    var menuRepository: MenuRepository,
    var storeRepository: StoreRepository,
    var userRepository: UserRepository,
    var storeMapper: StoreMapper
) {
    /**
     * 리뷰를 생성합니다.
     * @param storeId 가게 id
     * @param reviewFormDTO 리뷰 생성 요청 DTO
     * @param username 사용자 이름
     * @return 생성된 리뷰 정보
     */
    fun createReview(storeId: Long, reviewFormDTO: ReviewRequestDto, username: String): ReviewResponseDto {
        val reviewEntity = ReviewEntity(
            title = reviewFormDTO.title,
            content = reviewFormDTO.content,
            image = reviewFormDTO.image,
            score = reviewFormDTO.score,
            store = storeRepository.findById(storeId).get(),
            author = userRepository.findByEmail(username) ?: throw RuntimeException("가입되지 않은 유저입니다.")
        )
        return storeMapper.toResponseDto(reviewRepository.save(reviewEntity))
    }

    /**
     * 리뷰를 수정합니다.
     * @param reviewid 리뷰 id
     * @param reviewFormDTO 리뷰 수정 요청 DTO
     * @param username 사용자 이름
     * @return 수정된 리뷰 정보
     */
    fun updateReview(reviewid: Long, reviewFormDTO: ReviewRequestDto, username: String): ReviewResponseDto {
        val reviewEntity = reviewRepository.findById(reviewid).orElseThrow()
        reviewEntity.title = reviewFormDTO.title
        reviewEntity.content = reviewFormDTO.content
        reviewEntity.image = reviewFormDTO.image
        reviewEntity.score = reviewFormDTO.score
        return storeMapper.toResponseDto(reviewRepository.save(reviewEntity))
    }

    /**
     * 내 리뷰를 조회합니다.
     * @param username 사용자 이름
     * @return 내 리뷰 목록
     */
    fun getMyReview(username: String): List<ReviewResponseDto> {
        val userEntity = userRepository.findByEmail(username)  ?: throw RuntimeException("가입되지 않은 유저입니다.")
        return userEntity.review.map { reviewEntity ->
            storeMapper.toResponseDto(reviewEntity)
        }.toList()
    }

    /**
     * 리뷰를 삭제합니다.
     * @param reviewId 리뷰 id
     */
    @Transactional
    fun deleteReview(reviewId: Long) {
        reviewRepository.deleteById(reviewId)
    }
}
